packages/react/src/ReactServerContext.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {
+  REACT_PROVIDER_TYPE,
+  REACT_SERVER_CONTEXT_TYPE,
+  REACT_SERVER_CONTEXT_DEFAULT_VALUE_NOT_LOADED,
+} from 'shared/ReactSymbols';
+
+import type {
+  ReactServerContext,
+  ServerContextJSONValue,
+} from 'shared/ReactTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+
+import {enableServerContext} from 'shared/ReactFeatureFlags';
+
+const ContextRegistry = ReactSharedInternals.ContextRegistry;
+
+export function createServerContext<T: ServerContextJSONValue>(
+  globalName: string,
+  defaultValue: T,
+): ReactServerContext<T> {
+  if (!enableServerContext) {
+    throw new Error('Not implemented.');
+  }
+  let context;
+  if (ContextRegistry[globalName]) {
+    context = ContextRegistry[globalName];
+    if (
+      context._defaultValue === REACT_SERVER_CONTEXT_DEFAULT_VALUE_NOT_LOADED
+    ) {
+      context._defaultValue = defaultValue;
+    } else {
+      throw new Error(`ServerContext: ${globalName} already defined`);
+    }
+  } else {
+    context = ContextRegistry[globalName] = _createServerContext(
+      globalName,
+      defaultValue,
+    );
+  }
+  return context;
+}
+
+function _createServerContext<T: ServerContextJSONValue>(
+  globalName: string,
+  defaultValue: T,
+): ReactServerContext<T> {
+  const context: ReactServerContext<T> = {
+    $$typeof: REACT_SERVER_CONTEXT_TYPE,
+    // As a workaround to support multiple concurrent renderers, we categorize
+    // some renderers as primary and others as secondary. We only expect
+    // there to be two concurrent renderers at most: React Native (primary) and
+    // Fabric (secondary); React DOM (primary) and React ART (secondary).
+    // Secondary renderers store their context values on separate fields.
+    _currentValue: defaultValue,
+    _currentValue2: defaultValue,
+
+    _defaultValue: defaultValue,
+
+    // Used to track how many concurrent renderers this context currently
+    // supports within in a single renderer. Such as parallel server rendering.
+    _threadCount: 0,
+    _definitionLoaded: false,
+    // These are circular
+    Provider: (null: any),
+    displayName: globalName,
```

Comment:
This is not quite the same thing.

displayName is something you'd set for display nicely in DEV tools and we might not necessarily support that particular option longer term (given that it's deprecated even for functions now).

Where as the global name is something globally unique with ugly namespaces etc.

It's neat to reuse the property name but that property isn't actually always even available on contexts. Only if someone names them. (Which probably messes with hidden classes today.)

So it probably makes sense to store this separately.